### DISCUSSION TITLE
Optimal Integration of Opa with a REST API in NestJS: Experiences and Best Practices

### DISCUSSION THREAD

**Discussion by YupaTT** [open-policy-agent/community#564] [Community]
# Hello Opa community!

I'm excited to share with you a project I'm working on: integrating Opa with a REST API created in NestJS. During my research, I've come across a variety of approaches, but I'd like to tap into the collective experience of this community to identify best practices and optimal approaches.

To kick off the conversation, I'd like to share some of the resources I've been reading:

- **Official Opa Documentation:** As a starting point, I've reviewed the official Opa documentation to understand its architecture and recommended best practices.
- **Articles and tutorials on integrating Opa with other frameworks:** I've explored some online resources that address integrating Opa with other frameworks similar to NestJS, such as Express, to gain insights into how to approach this integration.
- **Feedback and posts in development forums:** Additionally, I've searched development forums and online communities to see the experiences of other developers who have worked on similar projects.

After this initial research, I've come up with some questions and considerations:

1. Does anyone in the community have direct experience integrating Opa with NestJS or a similar framework? I'd love to hear your experiences and challenges faced.
2. Are there specific design patterns you recommend to ensure a robust and scalable integration?
3. Is there any specific library or tool you would recommend to simplify the integration process?

I hope this discussion provides us with the opportunity to share knowledge and collaborate on identifying the best way to integrate Opa with a REST API in NestJS. I look forward to your contributions and experiences!

**Comment by srenatus** [comment]
Hello!

As luck would have it, I've been spending some time lately integrating OPA with NestJS, too. The result is [this example project](https://github.com/StyraInc/opa-typescript-example-nestjs), which makes use of the (also new) [opa-typescript SDK](https://github.com/StyraInc/opa-typescript/). I wouldn't go as far as claim that the NestJS integration done there is a "best practice", it's quite basic at this point, but I've been using it as real-world experiment to understand how to make the SDK more usable. Learning from it, we might come up with a higher level package targetting the use of OPA from NestJS.

In a gist, this is what it looks like at the moment:
```ts
@Controller('cats')
@AuthzQuery('cats/allow')
export class CatsController {
  constructor(private catsService: CatsService) {}

  @Post()
  @Authz({ action: 'create', resource: 'cat' })
  async create(@Body() createCatDto: CreateCatDto) {
    this.catsService.create(createCatDto);
  }
```
[link](https://github.com/StyraInc/opa-typescript-example-nestjs/blob/70620f9c60f2c8c2e4bdfa4707758445ab1d319e/src/cats/cats.controller.ts#L8-L16)

You're welcome to have a look, and to inform the discussion with your own use case(s)! Feel free to reach out on slack, too 👋
